src/groups/semidirect_product_group.jl

```diff
@@ -88,296 +96,885 @@ See [HilgertNeeb:2012; Definition 9.2.22](@cite), first definition for more deta
 
-Compute the group operation ``$(_math(:∘))`` on the semidirect product Lie group ``L = G ⋉ H``,
-that is for `g` `` = (g_1,h_1)``, `h` ``= (g_2,h_2)`` with ``g_1,g_2 ∈ G``, ``h_1,h_2 ∈ H``
-this computes
+$(_doc_semidirect_sub_groups) Let ``τ`` denote a right group action. It here acts on the left.
+
+The [`LeftSemidirectProductGroupOperation`](@ref) ``$(_math(:∘))`` on ``G ⋉ H`` is given by
+
+```math
+    (g_1,h_1) ∘ (g_2,h_2) := $(_tex(:bigl))( g_1 ⋆ g_2, τ_{g_2^{-1}}(h_1) ⋄ h_2 $(_tex(:bigr))).
+```
+
+The [`RightSemidirectProductGroupOperation`](@ref) ``$(_math(:∘))`` on ``H ⋊ G`` is given by
 
 ```math
-    (g_1,h_1) ∘ (g_2,h_2) := (g_1 ⋄ g_2, h_1 ⋆ σ_{g_1}(h_2)).
+    (h_1,g_1) ∘ (h_2,g_2) := $(_tex(:bigl))( τ_{g_2^{-1}}(h_1) ⋄ h_2, g_1 ⋆ g_2 $(_tex(:bigr))).
 ```
-where ``∘`` denotes the group operation on ``L``, ``⋄`` and ``⋆`` those on ``G`` and ``H``,
-respectively, and ``σ`` is the group action specified by the [`AbstractGroupActionType`](#ref) within the [`LeftSemidirectProductLieGroup`](@ref)  ``L``.
+
+See also [`AbstractRightGroupActionType`](@ref) and [`ActionActsOnLeft`](@ref).
 """
 compose(
-    SDPG::LieGroup{𝔽, LeftSemidirectProductGroupOperation, <:ProductManifold}, ::Any, ::Any
-) where {𝔽}
+    SDPG::LieGroup{𝔽, <:SemidirectProductGroupOperation{O1, O2, A, AO}, <:ProductManifold}, ::Any, ::Any
+) where {𝔽, O1, O2, A <: AbstractRightGroupActionType, AO <: ActionActsOnLeft}
 
+# An implementation for 1,3 (no inverse for left) and 5,7 (inverse for right)
 function _compose!(
-        SDPG::LieGroup{𝔽, <:LeftSemidirectProductGroupOperation, <:ProductManifold}, k, g, h
-    ) where {𝔽}
-    PM = SDPG.manifold
-    G, H = map(LieGroup, PM.manifolds, SDPG.op.operations)
-    A = GroupAction(SDPG.op.action_type, G, H)
-    # We have to perform 3 steps applying the group action
-    # 1) x = σ_g[1](h[2]) (a point on H)
-    # 2) compose g[1] and h[1] (a point on G, neither input needed afterwards)
-    # 3) compose g[2] and x (a point on H, neither input needed afterwards)
-    # to avoid to overwrite elements in case k=g or k=h: allocate for result of (1)
-    # especially after (1) we still need g[2] (in case k=g)
-    x = copy(H, submanifold_component(SDPG, k, Val(2)))
-    # (1)
-    apply!(
-        A, x, submanifold_component(SDPG, g, Val(1)), submanifold_component(SDPG, h, Val(2))
-    )
-    # (2)
-    _compose!(
-        G,
-        submanifold_component(SDPG, k, Val(1)),
-        submanifold_component(SDPG, g, Val(1)),
-        submanifold_component(SDPG, h, Val(1)),
-    )
-    _compose!(
-        H, submanifold_component(SDPG, k, Val(2)), submanifold_component(SDPG, g, Val(2)), x
-    )
+        SDPG::LieGroup{𝔽, <:SemidirectProductGroupOperation{O1, O2, A, AO}, <:ProductManifold}, k, g, h
+    ) where {𝔽, O1, O2, A <: AbstractGroupActionType, AO <: ActionActsOnLeft}
+    PM, G, H, a, g_ind, h_ind = _semidirect_parts(SDPG)
+    if Base.mightalias(k, g) || Base.mightalias(h, g) # k/h may not overlap with g
```

Comment:
```suggestion
    if Base.mightalias(k, g) || Base.mightalias(h, g) || Base.mightalias(k, h) # k/h may not overlap with g or k with h
```

Maybe a fix ... see next comment

Reply:
see #68 for tests that failed without this change.

Reply:
Hm, I sat down yesterday with pen&paper, drew even memory boxes nd had some energy drinks, but you are right I missed this, it seems.
I wanted to be careful to really only allocate when necessary. 

I am not at my computer, can you check, whether `mightalias` also exists with more than 2 arguments? Then we could replace this with one call. What about line 311? Is that fine then?

And - if we change this here we have to also check below 385ish? Maybe also there it is ... ah yes in 399 we have the same situation, so also that needs to be fixed. 

Could you do that?

Reply:
> I am not at my computer, can you check, whether mightalias also exists with more than 2 arguments?

Doesn't look like it.

> Could you do that?

I'll try ... this one already broke my brain and I was only able to find it because of a failed test.

Reply:
Thanks! And no worries, my brain as well yesterday, and I still did not get it right, and just doing the same edits to the other action side below (about line 385) as well then.

Reply:
Strangely the other action side didn't break the tests, but I'll have a look.

Reply:
Sure. In my head those two are symmetric, so should have the same memory structure, but it might also just be un my head ;)

Reply:
and currently it seems some of the 8 cases got lost in the merge? (or it is a hiccup of codecov)

Reply:
It looks like codecov hiccup, I did try and cover `diff_apply, diff_left_compose, diff_right_compose` but got errors and don't know if it is from me or bugs.

Reply:
No but then it is not a hiccup, we have to cover those.

But I now understand, your code cove report was green/ok because it was not worse than this PR (and it did not compare to master like this one).